File: 프로그래머스/2/389480. 완전범죄/완전범죄.py
def cache(func):
    mem = {}
    def wrapper(*args):
        key = tuple(args)
        if key in mem:
            return mem[key]
        res = func(*args)
        mem[key] = res
        return res
    return wrapper
        

def solution(info, n, m):
    max_size_n = 10e9
    info_len = len(info)

    @cache
    def get_min_a(i, prev):
        a, b = prev
        if a >= n or b >= m:
            return max_size_n
       	if i >= info_len:
            return a
        w_a, w_b = info[i]
        res_a = get_min_a(i+1, (a, b+w_b))
        res_b = get_min_a(i+1, (a+w_a, b))
        return min(res_a, res_b)
        
        
    res = get_min_a(0, (0, 0))
    return res if res != max_size_n else -1

File: 프로그래머스/2/389480. 완전범죄/완전범죄.js
function solution(info, n, m) {
  const len = info.length;
  
  // dp[i][a] = i번째 물건까지 처리했을 때, A가 a개 흔적을 남겼을 때 B의 최소 흔적
  const dp = Array(len + 1).fill(0)
    .map(() => Array(n).fill(Infinity));
  
  // 초기값: 0개 물건, A=0, B=0
  dp[0][0] = 0;
  
  // 각 물건에 대해
  for (let i = 0; i < len; i++) {
    for (let a = 0; a < n; a++) {
      if (dp[i][a] === Infinity) continue;
      
      const b = dp[i][a]; // 현재 B의 흔적
      
      // 선택 1: A가 i번째 물건을 훔침
      const nextA = a + info[i][0];
      if (nextA < n) { // A가 안 잡히는 범위
        dp[i + 1][nextA] = Math.min(dp[i + 1][nextA], b);
      }
      
      // 선택 2: B가 i번째 물건을 훔침
      const nextB = b + info[i][1];
      if (nextB < m) { // B가 안 잡히는 범위
        dp[i + 1][a] = Math.min(dp[i + 1][a], nextB);
      }
    }
  }
  
  // 모든 물건을 훔친 후, A의 최소 흔적 찾기
  let answer = -1;
  for (let a = 0; a < n; a++) {
    if (dp[len][a] < m) { // B도 안 잡히는 경우
      answer = a;
      break; // a가 작은 순서대로 확인하므로 첫 번째가 최소
    }
  }
  
  return answer;
}